src/nvim/ui.c

```diff
@@ -350,23 +307,33 @@ void ui_stop_highlight(void)
   set_highlight_args(current_attr_code);
 }
 
-void ui_visual_bell(void)
-{
-  UI_CALL(visual_bell);
-}
-
 void ui_puts(uint8_t *str)
 {
-  uint8_t *ptr = str;
+  uint8_t *p = str;
   uint8_t c;
 
-  while ((c = *ptr)) {
-    if (c < 0x20) {
-      parse_control_character(c);
-      ptr++;
-    } else {
-      send_output(&ptr);
+  while ((c = *p)) {
+    // TODO(bfredl): unbreak os/shell.c instead
```

Comment:
In https://github.com/neovim/neovim/pull/4376/files#diff-eba18ba3faa53d7aa0d8964997c5f015R308 I added a boolean to ui_puts, though I like your direction much better.